# Lecture 1 Exercises - Gym Tutorial

## Running an environment

In [11]:
import gym
env = gym.make('CartPole-v0')
# CartPole-v0, MountainCar-v0, 
env.reset()
for _ in range(1000):
    env.render()
    env.step(env.action_space.sample()) # take a random action

[2016-11-09 14:44:08,064] Making new env: CartPole-v0
[2016-11-09 14:44:08,204] You are calling 'step()' even though this environment has already returned done = True. You should always call 'reset()' once you receive 'done = True' -- any further steps are undefined behavior.


## Observations

In [12]:
import gym
env = gym.make('CartPole-v0')
for i_episode in range(20):
    observation = env.reset()
    for t in range(100):
        env.render()
        print(observation)
        action = env.action_space.sample()
        observation, reward, done, info = env.step(action)
        if done:
            print("Episode finished after {} timesteps".format(t+1))
            break

[2016-11-09 15:05:49,512] Making new env: CartPole-v0


[ 0.04009633 -0.00180397 -0.01438055 -0.01666391]
[ 0.04006025  0.19352123 -0.01471383 -0.31384916]
[ 0.04393067 -0.00138806 -0.02099081 -0.02584258]
[ 0.04390291 -0.1962028  -0.02150767  0.2601443 ]
[ 0.03997886 -0.00078052 -0.01630478 -0.03924408]
[ 0.03996325  0.1945714  -0.01708966 -0.3370264 ]
[ 0.04385467 -0.00030324 -0.02383019 -0.04978125]
[ 0.04384861 -0.19507553 -0.02482581  0.23528882]
[ 0.0399471   0.00039216 -0.02012004 -0.06512041]
[ 0.03995494  0.19579672 -0.02142245 -0.3640828 ]
[ 0.04387088  0.39121647 -0.0287041  -0.66344289]
[ 0.0516952   0.58672575 -0.04197296 -0.96502385]
[ 0.06342972  0.39219198 -0.06127344 -0.68581663]
[ 0.07127356  0.58810863 -0.07498977 -0.99714323]
[ 0.08303573  0.39406514 -0.09493263 -0.72892233]
[ 0.09091703  0.5903622  -0.10951108 -1.04991042]
[ 0.10272428  0.39684987 -0.13050929 -0.79351267]
[ 0.11066128  0.2037378  -0.14637954 -0.54456888]
[ 0.11473603  0.0109432  -0.15727092 -0.30135177]
[ 0.1149549  -0.18162849 -0.16329795 -0.06210648]


Inspect under the hood env.step()

In [15]:
env = gym.make('CartPole-v0')
env.reset()
for t in range(5):
    env.render()
    action = env.action_space.sample()
    observation, reward, done, info = env.step(action)
    print("observation: ", observation)
    print("reward: ", reward)
    print("done: ", done)
    print("info: ", info)
    if done:
        print("Episode finished after {} timesteps".format(t+1))
        break



[2016-11-09 15:11:19,648] Making new env: CartPole-v0


observation:  [-0.02995098  0.22697426  0.0264414  -0.30237673]
reward:  1.0
done:  False
info:  {}
observation:  [-0.0254115   0.42170958  0.02039386 -0.58660473]
reward:  1.0
done:  False
info:  {}
observation:  [-0.0169773   0.61654005  0.00866177 -0.87279429]
reward:  1.0
done:  False
info:  {}
observation:  [-0.0046465   0.42130139 -0.00879412 -0.57740076]
reward:  1.0
done:  False
info:  {}
observation:  [ 0.00377952  0.2263038  -0.02034213 -0.28750112]
reward:  1.0
done:  False
info:  {}


### Q: what does observation stand for?
observation:  [ 0.00377952  0.2263038  -0.02034213 -0.28750112]

What are these four numbers? Not in the documentation?


## Spaces

In [16]:
env = gym.make('CartPole-v0')
print(env.action_space)
#> Discrete(2)
print(env.observation_space)
#> Box(4,)


[2016-11-09 15:13:03,426] Making new env: CartPole-v0


Discrete(2)
Box(4,)


action_space methods

In [22]:
dir(env.observation_space)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'contains',
 'from_jsonable',
 'high',
 'low',
 'sample',
 'shape',
 'to_jsonable']

### Figure out how to interpret the action and observation space of CartPole-v0

Make 10 continous steps with action 0. See how observations change.

In [31]:
env = gym.make('CartPole-v0')
observation = env.reset()
print("First observation: ", observation)
for t in range(10):
    env.render()
    action = 0
    observation, reward, done, info = env.step(action)
    print("observation: ", observation)
    print("reward: ", reward)
    print("done: ", done)
    print("info: ", info)
    if done:
        print("Episode finished after {} timesteps".format(t+1))
        break

[2016-11-09 15:43:44,675] Making new env: CartPole-v0


First observation:  [ 0.02643812  0.00795602  0.01435272  0.03076028]
observation:  [ 0.02659724 -0.18736879  0.01496793  0.32793686]
reward:  1.0
done:  False
info:  {}
observation:  [ 0.02284987 -0.3827006   0.02152666  0.62530216]
reward:  1.0
done:  False
info:  {}
observation:  [ 0.01519585 -0.57811634  0.03403271  0.92468621]
reward:  1.0
done:  False
info:  {}
observation:  [ 0.00363353 -0.77368104  0.05252643  1.22786708]
reward:  1.0
done:  False
info:  {}
observation:  [-0.01184009 -0.96943821  0.07708377  1.53653352]
reward:  1.0
done:  False
info:  {}
observation:  [-0.03122886 -1.16539887  0.10781444  1.85224186]
reward:  1.0
done:  False
info:  {}
observation:  [-0.05453684 -1.36152875  0.14485928  2.17636455]
reward:  1.0
done:  False
info:  {}
observation:  [-0.08176741 -1.55773303  0.18838657  2.51002831]
reward:  1.0
done:  False
info:  {}
observation:  [-0.11292207 -1.75383891  0.23858714  2.8540414 ]
reward:  1.0
done:  True
info:  {}
Episode finished after 9 timest

A combination of experimenation and reading the source code. It seems:
* observation = [x, x_dot, theta, theta_dot]
* x is the horizontal position of the cart (right is positive).
* x_dot is the spead of x.
* theta is the angle between the pole and horizontal (clockwise is positive)

## Recording and uploading results

In [5]:
import gym
env = gym.make('CartPole-v0')
env.monitor.start('/Users/tongwang/Playground/rl_project_nov_2016/tmp/cartpole-experiment-1')
for i_episode in range(20):
    observation = env.reset()
    for t in range(100):
        env.render()
        print(observation)
        action = env.action_space.sample()
        observation, reward, done, info = env.step(action)
        if done:
            print("Episode finished after {} timesteps".format(t+1))
            break
env.monitor.close()

[2016-11-09 16:11:54,012] Making new env: CartPole-v0
[2016-11-09 16:11:54,019] Creating monitor directory /Users/tongwang/Playground/rl_project_nov_2016/tmp/cartpole-experiment-1
[2016-11-09 16:11:54,020] Starting new video recorder writing to /Users/tongwang/Playground/rl_project_nov_2016/tmp/cartpole-experiment-1/openaigym.video.3.56055.video000000.mp4


[-0.03940922  0.03992774  0.03605105  0.04066702]
[-0.03861066  0.23451467  0.03686439 -0.24042702]
[-0.03392037  0.42909115  0.03205585 -0.52125776]
[-0.02533855  0.62374755  0.0216307  -0.80366956]
[-0.0128636   0.4283358   0.0055573  -0.50426157]
[-0.00429688  0.62337899 -0.00452793 -0.795188  ]
[ 0.0081707   0.42831948 -0.02043169 -0.50393293]
[ 0.01673709  0.62372334 -0.03051035 -0.80298404]
[ 0.02921156  0.81925008 -0.04657003 -1.1051063 ]
[ 0.04559656  0.62477042 -0.06867215 -0.82738973]
[ 0.05809197  0.43065128 -0.08521995 -0.55707109]
[ 0.06670499  0.23682258 -0.09636137 -0.2924075 ]
[ 0.07144144  0.43317693 -0.10220952 -0.61385907]
[ 0.08010498  0.62956742 -0.1144867  -0.93690471]
[ 0.09269633  0.43615992 -0.13322479 -0.68227826]
[ 0.10141953  0.24311498 -0.14687036 -0.43432913]
[ 0.10628183  0.05034435 -0.15555694 -0.19131107]
[ 0.10728872 -0.14224977 -0.15938316  0.04854234]
[ 0.10444372 -0.33476998 -0.15841232  0.28700197]
[ 0.09774832 -0.52731996 -0.15267228  0.52583187]


[2016-11-09 16:11:54,551] Starting new video recorder writing to /Users/tongwang/Playground/rl_project_nov_2016/tmp/cartpole-experiment-1/openaigym.video.3.56055.video000001.mp4


[-0.00765882 -0.03214972  0.00412997  0.04411314]
[-0.00830182 -0.22733065  0.00501223  0.33809625]
[-0.01284843 -0.42252357  0.01177416  0.63235553]
[-0.0212989  -0.61780779  0.02442127  0.92872308]
[-0.03365506 -0.42302389  0.04299573  0.64381349]
[-0.04211553 -0.22852671  0.055872    0.3649742 ]
[-0.04668607 -0.03424145  0.06317148  0.09041889]
[-0.0473709  -0.23020932  0.06497986  0.40234442]
[-0.05197508 -0.42618983  0.07302675  0.71478542]
[-0.06049888 -0.23215069  0.08732246  0.44595324]
[-0.06514189 -0.42839248  0.09624152  0.76483454]
[-0.07370974 -0.23471828  0.11153821  0.50391897]
[-0.07840411 -0.04133045  0.12161659  0.24836407]
[-0.07923072 -0.23796036  0.12658388  0.57679762]
[-0.08398993 -0.43460807  0.13811983  0.90652545]
[-0.09268209 -0.63130256  0.15625034  1.23923563]
[-0.10530814 -0.82804731  0.18103505  1.57650846]
Episode finished after 17 timesteps


[2016-11-09 16:11:54,918] Starting new video recorder writing to /Users/tongwang/Playground/rl_project_nov_2016/tmp/cartpole-experiment-1/openaigym.video.3.56055.video000008.mp4


[  2.98647413e-02  -2.51179127e-02  -4.47179360e-05  -2.96725596e-02]
[ 0.02936238 -0.22023922 -0.00063817  0.26299626]
[ 0.0249576  -0.41535206  0.00462176  0.55547783]
[ 0.01665056 -0.6105386   0.01573131  0.8496133 ]
[ 0.00443979 -0.41563467  0.03272358  0.5619184 ]
[-0.00387291 -0.61120019  0.04396195  0.86472865]
[-0.01609691 -0.41670333  0.06125652  0.58618589]
[-0.02443098 -0.61262735  0.07298024  0.89751885]
[-0.03668353 -0.41856652  0.09093061  0.62863961]
[-0.04505486 -0.61483195  0.10350341  0.94851827]
[-0.05735149 -0.81118366  0.12247377  1.27184531]
[-0.07357517 -1.00763721  0.14791068  1.60023566]
[-0.09372791 -0.81454439  0.17991539  1.35708631]
[-0.1100188  -1.01140848  0.20705712  1.70022624]
Episode finished after 14 timesteps
[ 0.01250093 -0.03953976 -0.02556801  0.04387704]
[ 0.01171013  0.15593931 -0.02469047 -0.25676192]
[ 0.01482892 -0.03882159 -0.02982571  0.02803212]
[ 0.01405249 -0.2335034  -0.02926507  0.31115751]
[ 0.00938242 -0.42819645 -0.02304192  0.5944

[2016-11-09 16:11:55,544] Finished writing results. You can upload them to the scoreboard via gym.upload('/Users/tongwang/Playground/rl_project_nov_2016/tmp/cartpole-experiment-1')


[-0.00471712  0.02472492 -0.00586879  0.00373597]
[-0.00422263 -0.17031237 -0.00579407  0.29456146]
[ -7.62887384e-03  -3.65351241e-01   9.71569807e-05   5.85411403e-01]
[-0.0149359  -0.17023065  0.01180539  0.29275908]
[-0.01834051  0.02472101  0.01766057  0.00382267]
[-0.01784609  0.21958529  0.01773702 -0.28323623]
[-0.01345439  0.0242149   0.0120723   0.01498767]
[-0.01297009  0.21916166  0.01237205 -0.27386199]
[-0.00858685  0.41410492  0.00689481 -0.56261721]
[-0.00030476  0.21888689 -0.00435754 -0.26777008]
[ 0.00407298  0.0238274  -0.00971294  0.02353527]
[ 0.00454953  0.21908729 -0.00924223 -0.2721963 ]
[ 0.00893128  0.41433989 -0.01468616 -0.56777988]
[ 0.01721807  0.60966473 -0.02604175 -0.86505312]
[ 0.02941137  0.80513128 -0.04334282 -1.16580894]
[ 0.04551399  0.61059943 -0.066659   -0.88702399]
[ 0.05772598  0.80655979 -0.08439948 -1.19989496]
[ 0.07385718  1.00266603 -0.10839737 -1.51779125]
[ 0.0939105  0.8090092 -0.1387532 -1.2608174]
[ 0.11009068  0.61590752 -0.163969